app/models/external_report.rb

```diff
@@ -18,42 +17,31 @@ def options_for_report_type
     ReportTypes.map { |rt| [rt, rt] }
   end
 
-  def offering_api_url(offering_id, protocol, host)
-    routes = Rails.application.routes.url_helpers
-    opts = {protocol:protocol, host:host}
-    case report_type
-      when OfferingReport
-        routes.api_v1_offering_url(offering_id, opts)
-      when ClassReport
-        routes.for_class_api_v1_offering_url(offering_id, opts)
-      when TeacherReport
-        routes.for_teacher_api_v1_offering_url(offering_id, opts)
-      else
-        routes.api_v1_offering_url(offering_id, opts)
-    end
-  end
-
-  # Return a the external_report url
-  # with parameters for the offering_api_url
-  # and the short-lived bearer token for the user.
-  def url_for(offering_id, user, protocol, host)
+  # Return a the external_report url and the short-lived bearer token for the user.
+  def url_for_offering(offering, user, protocol, host)
     grant = client.updated_grant_for(user, ReportTokenValidFor)
     if user.portal_teacher
       grant.teacher = user.portal_teacher
       grant.save!
     end
     token = grant.access_token
     username = user.login
-    "#{url}?offering=#{offering_api_url(offering_id, protocol, host)}&token=#{token}&username=#{username}"
+    routes = Rails.application.routes.url_helpers
+    offering_url = CGI.escape(routes.api_v1_offering_url(offering.id, protocol: protocol, host: host))
+    class_id = offering.clazz.id
+    class_offerings_url = CGI.escape(routes.api_v1_offerings_url(class_id: class_id, protocol: protocol, host: host))
+    class_url = CGI.escape(routes.api_v1_class_url(class_id, protocol: protocol, host: host))
+    "#{url}?reportType=offering&offering=#{offering_url}&classOfferings=#{class_offerings_url}&class=#{class_url}&token=#{token}&username=#{username}"
```

Comment:
The URL might already have parameters in it. So it would be better to check for that and not add a second `?` if so.  This code does this: https://github.com/concord-consortium/rigse/blob/6a1f27d36b149612ce90c407f6fd9c0b8bccee08/app/controllers/application_controller.rb#L294-L299 I believe it will remove the need for the `CGI.escape` This isn't a regression, so I'm not going to hold up the PR for this.